File: benchmarks/haskell/src/Concurrency/ParallelMap.hs
module Concurrency.ParallelMap
  ( parallelMapSum
  , sequentialMapSum
  ) where

import Control.Parallel.Strategies (parMap, rdeepseq)
import Data.Vector (Vector)
import qualified Data.Vector as V

-- | Parallel map: compute sqrt of each element using Strategies.
-- Returns the sum to force evaluation of all elements.
parallelMapSum :: Vector Double -> Double
parallelMapSum v =
    let xs = V.toList v
        results = parMap rdeepseq sqrt xs
     in sum results

-- | Sequential version for comparison.
sequentialMapSum :: Vector Double -> Double
sequentialMapSum = V.sum . V.map sqrt

File: rust/src/concurrency/parallel_map.rs
use rayon::prelude::*;

/// Parallel map: compute sqrt of each element using rayon.
/// Returns the sum to force evaluation of all elements.
pub fn parallel_map_sum(data: &[f64]) -> f64 {
    data.par_iter().map(|x| x.sqrt()).sum()
}

/// Sequential version for comparison.
pub fn sequential_map_sum(data: &[f64]) -> f64 {
    data.iter().map(|x| x.sqrt()).sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_produce_same_result_parallel_and_sequential() {
        let data: Vec<f64> = (1..=10000).map(|i| i as f64).collect();
        let par = parallel_map_sum(&data);
        let seq = sequential_map_sum(&data);
        assert!((par - seq).abs() < 1e-6);
    }
}
